pkg/engine/planner/logical/planner.go

```diff
@@ -0,0 +1,236 @@
+package logical
+
+import (
+	"fmt"
+
+	"github.com/prometheus/prometheus/model/labels"
+
+	"github.com/grafana/loki/v3/pkg/engine/internal/types"
+	"github.com/grafana/loki/v3/pkg/logproto"
+	"github.com/grafana/loki/v3/pkg/logql"
+	"github.com/grafana/loki/v3/pkg/logql/log"
+	"github.com/grafana/loki/v3/pkg/logql/syntax"
+)
+
+// BuildPlan converts a LogQL query represented as [logql.Params] into a logical [Plan].
+// It may return an error as second argument in case the traversal of the AST of the query fails.
+func BuildPlan(query logql.Params) (*Plan, error) {
+	var selector Value
+	var predicates []Value
```

Comment:
nit(future): handling queries with multiple leaf nodes where different legs have different selectors and predicates, e.g.
```
sum(rate({job="foo"} |= "bar" [5m))
/
sum(rate({job="bazz"} |= "buzz" [5m))
```

Might be considering something like [fold](https://en.wikipedia.org/wiki/Fold_(higher-order_function)) as an alternative to `expr.Walk`